crates/evm/evm/src/executors/invariant/error.rs

```diff
@@ -202,70 +224,137 @@ impl InvariantFuzzError {
+    /// After running either the powerset or the random sequences, we check if we successfully
+    /// shrunk the call sequence.
+    fn try_shrinking_recurse(
+        &self,
+        calls: &[BasicTxDetails],
+        executor: &Executor,
+        runs: usize,
+        retries: usize,
+    ) -> Vec<usize> {
+        let shrunk = Arc::new(RwLock::new((0..calls.len()).collect()));
+        let shrink_limit = self.shrink_run_limit - runs;
+
+        // we construct either a full powerset (this guarantees we maximally shrunk for the given
+        // calls) or a random subset
+        let (set, is_powerset): (Vec<_>, bool) = if calls.len() <= 64 &&
+            2_usize.pow(calls.len() as u32) <= shrink_limit
+        {
+            // we add the last tx always because thats ultimately what broke the invariant
+            let powerset = (0..calls.len() - 1)
+                .powerset()
+                .map(|mut subset| {
+                    subset.push(calls.len() - 1);
+                    subset
+                })
+                .collect();
+            (powerset, true)
+        } else {
+            let mut rng = thread_rng();
+            (
+                (0..shrink_limit / 3)
+                    .map(|_| {
+                        // select between 1 and calls.len() - 2 number of indices
+                        let amt: usize = rng.gen_range(1..calls.len() - 1);
+                        // construct a random sequence of indices, up to calls.len() - 1 (sample is
+                        // exclusive range and we dont include the last tx
+                        // because its always included), and amt number of indices
+                        let mut seq = seq::index::sample(&mut rng, calls.len() - 1, amt).into_vec();
+                        // sort the indices
+                        seq.sort();
+                        // we add the last tx always because thats what ultimately broke the
+                        // invariant
+                        seq.push(calls.len() - 1);
+                        seq
+                    })
+                    .collect(),
+                false,
+            )
+        };
 
-                removed_calls.push(next_removed);
-            } else {
-                // When the process is restarted, `removed_calls` will be empty.
-                removed_calls.push(anchor + 1);
-            }
+        let new_runs = set.len();
+
+        // just try all of them in parallel
+        set.par_iter().for_each(|use_calls| {
+            self.set_fails_successfully(executor.clone(), calls, use_calls, Arc::clone(&shrunk));
+        });
+
+        // at this point we know shrunk is fine to get the inner value for
+        let shrunk = Arc::<RwLock<Vec<usize>>>::try_unwrap(shrunk).unwrap().into_inner();
+
+        if is_powerset {
+            // a powerset is guaranteed to be smallest subset
+            shrunk
+        } else if shrunk.len() < calls.len() && new_runs + runs < self.shrink_run_limit {
```

Comment:
This means we can keep making progress because we haven't hit the shrink run limit right?

Reply:
it’s a check that we made progress shrinking and we still have computation budget to try to keep shrinking. will simplify